src/System.Net.Security/src/System/Net/SecureProtocols/_SslStream.cs

```diff
@@ -0,0 +1,709 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.IO;
+using System.Threading;
+
+namespace System.Net.Security
+{
+    //
+    // This is a wrapping stream that does data encryption/decryption based on a successfully authenticated SSPI context.
+    //
+    internal class _SslStream
+    {
+        private static AsyncCallback s_writeCallback = new AsyncCallback(WriteCallback);
+        private static AsyncProtocolCallback s_resumeAsyncWriteCallback = new AsyncProtocolCallback(ResumeAsyncWriteCallback);
+        private static AsyncProtocolCallback s_resumeAsyncReadCallback = new AsyncProtocolCallback(ResumeAsyncReadCallback);
+        private static AsyncProtocolCallback s_readHeaderCallback = new AsyncProtocolCallback(ReadHeaderCallback);
+        private static AsyncProtocolCallback s_readFrameCallback = new AsyncProtocolCallback(ReadFrameCallback);
+
+        private const int PinnableReadBufferSize = 4096 * 4 + 32;         // We read in 16K chunks + headers.
+        private static PinnableBufferCache s_PinnableReadBufferCache = new PinnableBufferCache("System.Net.SslStream", PinnableReadBufferSize);
+        private const int PinnableWriteBufferSize = 4096 + 1024;        // We write in 4K chunks + encryption overhead.
+        private static PinnableBufferCache s_PinnableWriteBufferCache = new PinnableBufferCache("System.Net.SslStream", PinnableWriteBufferSize);
+
+        private SslState _SslState;
+        private int _NestedWrite;
+        private int _NestedRead;
+
+        // Never updated directly, special properties are used.  This is the read buffer.
+        private byte[] _InternalBuffer;
+        private bool _InternalBufferFromPinnableCache;
+
+        private byte[] _PinnableOutputBuffer;                        // Used for writes when we can do it. 
+        private byte[] _PinnableOutputBufferInUse;                   // Remembers what UNENCRYPTED buffer is using _PinnableOutputBuffer.
+
+        private int _InternalOffset;
+        private int _InternalBufferCount;
+
+        private FixedSizeReader _Reader;
+
+        internal _SslStream(SslState sslState)
+        {
+            if (PinnableBufferCacheEventSource.Log.IsEnabled())
+            {
+                PinnableBufferCacheEventSource.Log.DebugMessage1("CTOR: In System.Net._SslStream.SslStream", this.GetHashCode());
+            }
+            _SslState = sslState;
+            _Reader = new FixedSizeReader(_SslState.InnerStream);
+        }
+
+        // If we have a read buffer from the pinnable cache, return it.
+        private void FreeReadBuffer()
+        {
+            if (_InternalBufferFromPinnableCache)
+            {
+                s_PinnableReadBufferCache.FreeBuffer(_InternalBuffer);
+                _InternalBufferFromPinnableCache = false;
+            }
+            _InternalBuffer = null;
+        }
+
+        ~_SslStream()
```

Comment:
As far as I can tell, every SslStream is going to put an _SslStream onto the finalizer queue, why isn't this type Disposable?